### ISSUE TITLE
Configurable Salt for AES for Deterministic Zips

### ISSUE THREAD

**Issue by Philip-Trettner** [zip-rs/zip2#223]
I'm building a reproducible/deterministic build/action system and some parts require password protected zip files.

Some builds were not reproducible and I could narrow it down to the password protected zips.

In particular, the AesWriter::new creates a `salt` using `rand::thread_rng()` (https://github.com/zip-rs/zip2/blob/master/src/aes.rs#L240).

As far as I understand it, this makes it impossible to have deterministic/reproducible zips with Aes as you cannot seed the thread_rng.

So my feature request for _one_ of the following:
* add an option to provide a custom Rng
* add an option to provide a custom Seed
* add an option to provide a custom Salt
* add an option to fall back to a fixed Salt
* or some other option that makes the process deterministic

(I basically need only a single way to get determinism here. The password itself is derived from a build hash, so there is no security implication in my case afaict.)

I'd be happy to make a PR if you're ok with such an option and provide me with a preferred way.

**Comment by im7mortal** [comment]
I believe it is possible to implement this as I mentioned in [ this comment](https://github.com/zip-rs/zip2/issues/170#issuecomment-3917300648).

Instead of providing multiple configuration options, the user could implement their own version of the encryption writer and inject it.

In that case, the user would need to copy the existing implementation, adjust the required logic, and then inject it into the `ZipWriter`.

https://github.com/zip-rs/zip2/blob/59e59b1c2407bb8c01dcae1ec9a4430ab70ca9a2/src/aes.rs#L225-L320

**Comment by Its-Just-Nans** [comment]
> In that case, the user would need to copy the existing implementation, adjust the required logic, and then inject it into the `ZipWriter`.


Yes, but if we want a simple solution, we can use the options i think. The user would not have to rewrite the whole implementation (just to have reproducible build).

**Comment by im7mortal** [comment]
In my opinion, we need to protect the library from features that could bloat it, introduce instability, and increase the surface area for errors.

This applies to all parts of the library, but especially to security-related modules — in this case, the crypto AES module. If we introduce these additional options, we will likely need to add conditional logic, validate provided values, and define new error types. Over time, this could increase complexity and potentially lead to more issues.

We should also assess how many users would benefit from this feature, as it may be a niche use case.

At the same time, there will always be requests for more specific features. As I mentioned in [this comment](https://github.com/zip-rs/zip2/issues/204#issuecomment-3917510922), the injection approach helps preserve a stable core implementation while still allowing flexibility.

**Comment by im7mortal** [comment]
:bangbang:

When I wrote the earlier commit, I didn’t yet know that a PR had already been created.

@Pr0methean @Its-Just-Nans  I should express my concerns more clearly.

1. As I predicted, this feature increases the complexity of the codebase.
2. AES is a security-sensitive module. This issue has been stale for two years, and it’s possible that nobody needs it anymore (@Philip-Trettner).
3. The AES module `new` function is currently well written and easy to review. This PR introduces additional `if let` conditions, which make the code harder to read and reason about:
https://github.com/zip-rs/zip2/blob/08f24726af11cc7beaf2ec6d1ac02e6b99319c6e/src/aes.rs#L316-L329
4.Introducing `AesModeOptions` sets a precedent. Other users may request similar feature additions. **If a user wants their own cryptography, they should implement it themselves.** That’s why I’ve been advocating for an injection-based approach.
5.Any changes to cryptographic modules (unless they are security hotfixes) should involve thorough discussion and a cooling-off period before implementation begins.

**Comment by Philip-Trettner** [comment]
It's still part of my workflow and I would appreciate determinism a lot. That being said, I agree with your concerns about increased complexity. I'm totally fine with an injection based approach and a small amount of code duplication on my side.

(For reference: I have a content-addressed-hashing based build pipeline written in rust where zipping & unzipping files is an important intermediate step. Think: I build and package my software (https://solidean.com/) and the finished artifact is then zipped. it is then immediately unzipped and the public facing test suite is run against it. This design makes sure that we test everything as close to the released artifacts as possible. The missing determinism means that if any step of the pipeline before the zipping is re-run, the zip will differ in bytes even if the input is the same. This causes spurious updates downstream. While not catastrophic, determinism would be nice here)

**Comment by Its-Just-Nans** [comment]
Hi @im7mortal 

Thanks for your feedback on the crate, here are my answer concerning the differents points.

Note, this is also related to https://github.com/zip-rs/zip2/pull/680 and https://github.com/zip-rs/zip2/pull/680#issuecomment-3919357143

> 1. As I predicted, this feature increases the complexity of the codebase.

Yes, it increases the complexity of the codebase. I think adding any code to the crate will inevitably increase the size (and potentially the complexity) of the codebase.

---

> 2. AES is a security-sensitive module. This issue has been stale for two years, and it’s possible that nobody needs it anymore ([@Philip-Trettner](https://github.com/Philip-Trettner)).

It seems like @Philip-Trettner is still interested in this feature. I also think this feature could be useful for additional testing of ZIP generation. However, it is most likely going to be useful for reproducible builds (since adding a non-random salt is generally not a great idea).

---

> 3. The AES module `new` function is currently well written and easy to review. This PR introduces additional `if let` conditions, which make the code harder to read and reason about:

It's a single `if let` on an `Option<>`, which is a quite common pattern. Since the variable is named `custom_salt`, I also think that it's not that hard to read or understand.
Note that some parts  of the code in this crate are much harder to read than this simple `if let`

---

>    4.Introducing `AesModeOptions` sets a precedent. Other users may request similar feature additions. **If a user wants their own cryptography, they should implement it themselves.** That’s why I’ve been advocating for an injection-based approach.

I agree that creating a custom cryptography could be an idea but I think we prefer for now to create zip files which respect the zip specifications.
Looking at your comment on this Issue https://github.com/zip-rs/zip2/issues/675

> > almost everything I check is actually outdated, not extendable, and not safe — so it probably shouldn’t be implemented at all.

So for now the needs for a new cryptography in the crate is questionable

---

>    5.Any changes to cryptographic modules (unless they are security hotfixes) should involve thorough discussion and a cooling-off period before implementation begins.

Yes I agree on this one. But technically on this change, we are not changing anythinh on the crypto side , and we are warning the user that they should use it at their own risk. The custom salt is not a default option; the library user needs to provide the correct `FileOption` (with `with_aes_encryption_and_salt` to have enable the custom salt.

End notes
- my PR could maybe be updated to not have a `mode` duplication.
- this crate has still some technical debt, and I'm not sur that taking a new path with `dyn trait objects` is the way to go

I’m waiting for @Pr0methean ’s response with interest.

**Comment by im7mortal** [comment]
Thank you for the detailed answer Its-Just-Nans.

I am very interested in hearing the Pr0methean's opinion. :smiley:

Regarding the issue, I feel the recent activity happened in a _a non-deterministic way_ :grin:. The issue had been stale for almost two years, which is why I decided to work on it. I did not expect my early investigation to trigger so much activity. I found it while working on another issue and exploring the library internals, looking for connections between the code and related issues.

I believe the approach discussed in #170 could fit here. However, it requires time to assess properly. I plan to benchmark it and provide more detailed reasoning. That is why I am asking for **a cool down on this implementation**.

**Comment by Pr0methean** [comment]
I agree with @Its-Just-Nans that the complexity cost of handling a user-provided salt is acceptable -- and not a reason for a delay -- as long as we don't break backward-compatibility for more basic usage. After all, it's much smaller than the cost of adding a new compression library, and the ZIP spec contains lots of compression methods that we still don't support and that someone's likely to need sooner or later.

**Comment by im7mortal** [comment]
@Pr0methean @Its-Just-Nans  

**Can we please put a one-month freeze on this implementation?**

This issue was inactive for two years before I started investigating it. During that time, no one requested the feature, and the original author has had a workaround in place. @Philip-Trettner 

My concern is not the specific implementation, but the process. After I mentioned that I had an alternative approach, a PR appeared within two hours, and now there is suddenly urgency around delivery.

It was the middle of the work week, and I didn’t have time to argue for a different approach. All I was asking for was time to work on this properly so I could provide my version for review and discussion.

**Comment by Pr0methean** [comment]
@Philip-Trettner How badly would it hurt if the fix weren't released or merged until March 23, and you had to declare a Git dependency on someone's fork to use it sooner, versus if it were merged and released this coming Monday, February 23? Because that's what @im7mortal is asking for, even before we consider the extra headaches for the team of members who'd to maintain both a "master" branch without the change and an *actual* trunk with it in the meantime.

Sorry if I sound unusually negative; but I've heard maintainers' opinions about backporting fixes to multiple versions of things like OpenJDK, and none of the opinions I've heard were positive. I also tried it with https://github.com/Pr0methean/BetterRandom/tree/java7 myself once upon a time, and my experience wasn't encouraging -- and that was when the users and collaborators were hypothetical ones that I only needed to pretend to have for self-training and portfolio-building purposes.

**Comment by Its-Just-Nans** [comment]
> **Can we please put a one-month freeze on this implementation?**

On my side that's **not a problem to wait and see your proposal** (also I need to do a tiny fix to the MR).
But note that we may not accept it (e.g. too complex, too much breaking changes, will introduce more issues than benefits, difficult to maintain...)

About the process and the timings of my pull request, yes I understand that that can be perceived as a fast change after your comment. Note that my pull request wasn’t entirely triggered by your activity on the issue. From time to time I check the old issues of the repo and see if we can fix/implement them and how easy it could be. Sometimes I do PRs, sometimes I don't have the time or I don't feel it because it can be complex (or it takes more time to do the fix). This issue was (and is) one of the issues that I put on my "think about it and do something a day" list. So I had already thought about it, but hadn't written actual code. When you commented on the issue I thought that I can produce a relatively easy and simple fix. Note my comment

> Yes, but if we want a simple solution...

So I tried for a simple solution to it and now we have a merge request (still needs tiny fix though). The overall idea of the pull request is here, and it's in the same way as this crate: with a new field/function on `FileOption`

**Comment by Pr0methean** [comment]
@im7mortal You can still make your proposal whenever it's ready; just start it by reverting the parts of Nans' change you can improve on.
